parser/charset/encoding_ascii.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package charset
+
+import (
+	go_unicode "unicode"
+
+	"golang.org/x/text/encoding"
+)
+
+// EncodingASCIIImpl is the instance of encodingASCII
+var EncodingASCIIImpl = &encodingASCII{encodingBase{enc: encoding.Nop}}
+
+func init() {
+	EncodingASCIIImpl.self = EncodingASCIIImpl
+}
+
+// encodingASCII is the ASCII encoding.
+type encodingASCII struct {
+	encodingBase
+}
+
+// Name implements Encoding interface.
+func (e *encodingASCII) Name() string {
+	return CharsetASCII
+}
+
+// Peek implements Encoding interface.
+func (e *encodingASCII) Peek(src []byte) []byte {
+	if len(src) == 0 {
+		return src
+	}
+	return src[:1]
+}
+
+func (e *encodingASCII) Transform(dest, src []byte, op Op) ([]byte, error) {
+	if IsValid(e, src) {
+		return src, nil
+	}
+	return e.encodingBase.Transform(dest, src, op)
+}
+
+func (e *encodingASCII) Foreach(src []byte, op Op, fn func(from, to []byte, ok bool) bool) {
+	for i, w := 0, 0; i < len(src); i += w {
+		w = 1
+		ok := true
+		if src[i] > go_unicode.MaxASCII {
+			w = len(EncodingUTF8Impl.Peek(src[i:]))
```

Comment:
why `EncodingUTF8Impl.Peek`?

Reply:
Because I am trying to make it compatible with MySQL:

```sql
set @@sql_mode = '';
drop table if exists t;
create table t (a char(10) charset ascii);
insert into t values ('😊');
select * from t;
```

```
mysql> select * from t;
+------+
| a    |
+------+
| ?    |
+------+
1 row in set (0.00 sec)
```

Note there is only one `?`.

Reply:
if set `character_set_client=ascii`
```sql
mysql> insert into t values ('😊');
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> select * from t;
+------+
| a    |
+------+
| ???? |
+------+
1 row in set (0.00 sec)
```